src/Engine/ProtoCore/Namespace/ElementRewriter.cs

```diff
@@ -0,0 +1,213 @@
+
+        private void LookupResolvedNameAndRewriteAst(ClassTable classTable, ref AssociativeNode astNode)
+        {
+            Debug.Assert(elementResolver != null);
+
+            // Get partial class identifier/identifier lists
+            var classIdentifiers = GetClassIdentifiers(astNode);
+
+            var resolvedNames = new Queue<string>();
+            foreach (var identifier in classIdentifiers)
+            {
+                var partialName = CoreUtils.GetIdentifierExceptMethodName(identifier);
+                var resolvedName = elementResolver.LookupResolvedName(partialName);
+                if (string.IsNullOrEmpty(resolvedName))
+                {
+                    // If namespace resolution map does not contain entry for partial name, 
+                    // back up on compiler to resolve the namespace from partial name
+                    var matchingClasses = CoreUtils.GetResolvedClassName(classTable, identifier);
+
+                    if (matchingClasses.Length == 1)
+                    {
+                        resolvedName = matchingClasses[0];
+                        var assemblyName = CoreUtils.GetAssemblyFromClassName(classTable, resolvedName);
+
+                        elementResolver.AddToResolutionMap(partialName, resolvedName, assemblyName);
+                    }
+                    else
+                    {
+                        // Namespace conflict, class name could not be resolved.
+                        // This will be reported subsequently in the pre-compilation stage
+                        return;
+                    }
+                }
+                resolvedNames.Enqueue(resolvedName);
+            }
+            
+            RewriteAstWithResolvedName(ref astNode, ref resolvedNames);
+        }
+
+        private static IEnumerable<IdentifierListNode> GetClassIdentifiers(AssociativeNode astNode)
+        {
+            var classIdentifiers = new List<IdentifierListNode>();
+            var resolvedNames = new Queue<string>();
+            DfsTraverse(ref astNode, ref classIdentifiers, ref resolvedNames);
+            return classIdentifiers;
+        }
+
+        /// <summary>
+        /// Replace partial identifier with fully resolved identifier list in original AST
+        /// This is the same AST that is also passed to the VM for execution
+        /// </summary>
+        /// <param name="astNode"></param>
+        /// <param name="resolvedNames"> fully qualified class identifier list </param>
+        private static void RewriteAstWithResolvedName(ref AssociativeNode astNode, ref Queue<string> resolvedNames)
+        {
+            List<IdentifierListNode> classIdentifiers = null;
+            DfsTraverse(ref astNode, ref classIdentifiers, ref resolvedNames);
+        }
+
+        #region private utility methods
+
+        private static void DfsTraverse(ref AssociativeNode astNode, ref List<IdentifierListNode> classIdentifiers, ref Queue<string> resolvedNames)
```

Comment:
I don't see why `resolvedNames` needs to be an argument by `ref`, if that's accurate, please remove the `ref` for it.